core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.headers
+
+import java.net.{Inet4Address, Inet6Address}
+
+import cats.data.NonEmptyList
+import org.http4s.parser.{ForwardedModelParsers, HttpHeaderParser}
+import org.http4s.util.Writer
+import org.http4s.{Header, HeaderKey, ParseResult, Uri}
+
+object Forwarded
+    extends HeaderKey.Internal[Forwarded]
+    with HeaderKey.Recurring
+    with ForwardedModelParsers {
+
+  final case class Node(nodeName: Node.Name, nodePort: Option[Node.Port] = None)
+
+  object Node {
+    def apply(nodeName: Name, nodePort: Port): Node = apply(nodeName, Some(nodePort))
+
+    sealed trait Name extends Any
+
+    object Name {
+      final case class Ipv4(address: Uri.Ipv4Address) extends AnyVal with Name
+      final case class Ipv6(address: Uri.Ipv6Address) extends AnyVal with Name
+      case object Unknown extends Name
+
+      def apply(address: Uri.Ipv4Address): Name = Name.Ipv4(address)
+      def apply(address: Inet4Address): Name = apply(Uri.Ipv4Address.fromInet4Address(address))
+      def apply(a: Byte, b: Byte, c: Byte, d: Byte): Name = apply(Uri.Ipv4Address(a, b, c, d))
+
+      def apply(address: Uri.Ipv6Address): Name = Name.Ipv6(address)
+      def apply(address: Inet6Address): Name = apply(Uri.Ipv6Address.fromInet6Address(address))
+      def apply(a: Short, b: Short, c: Short, d: Short, e: Short, f: Short, g: Short, h: Short)
+          : Name = apply(Uri.Ipv6Address(a, b, c, d, e, f, g, h))
+    }
+
+    sealed trait Port extends Any
+
+    object Port {
+      final case class Num(value: Int) extends AnyVal with Port
```

Comment:
If we're being real strict about the spec, only values 0 to 99999 are allowed.  (Sadly, not 0 to 65535, or we could wrap a `Short`!)

Reply:
Well, I could make its ctor private and add something like `fromInt` method to `Port` if you think it might make sense.

Reply:
`fromInt(port: Int): Either[ParseError, Port]` might make sense

Reply:
@rossabaker I wonder how come the port range you mentioned is 0..99999? I looked through a bunch of specs but it seems that everywhere it is 0..65535, even for IPv6.

Reply:
@hamnis if we would like to guarantee correctness of the port number in `Forwarded.Node.Port` instances, then perhaps it makes sense to do the same for `Forwarded.Host`. But I'm struggling to come up with some nice name for such constructor, because it takes two parameters – `Uri.Host` and `Int` simultaneously. Would something like
```
object Host {
  def parse(uriHost: Uri.Host, portNum: Int): ParseResult[Host]
}
```
be good enough or could you suggest some nicer naming?

Reply:
`parse` does not really sit well with me here.
what about just `from` 

maybe a constructor from `Uri` might make sense as well.

`def fromUri(uri: Uri): ParseResult[Host]`